pkg/cloud/services/compute/instance.go

```diff
@@ -508,10 +504,22 @@ func (s *Service) getOrCreatePort(eventObject runtime.Object, clusterName string
 	}
 
 	record.Eventf(eventObject, "SuccessfulCreatePort", "Created port %s with id %s", port.Name, port.ID)
+
+	if portOpts.Trunk {
```

Comment:
This is a problem.

Previously if we specified trunk on the instance and a number of ports without any trunk setting, those ports would inherit the trunk setting of the instance. Here we're *always* using the trunk setting of the port, which now defaults to False rather than inheriting the trunk setting of the instance. This will cause a change of behaviour.

I think we want the following behaviour:
1. instance=true, port=undefined, trunk=true
2. instance=true, port=true, trunk=true
3. instance=true, port=false, trunk=false
4. instance=false, port=undefined, trunk=false
5. instance=false, port=true, trunk=true
6. instance=false, port=false, trunk=false

If we define port=undefined to be false (the behaviour here), that breaks case 1.
If we define port=undefined to be true, that breaks case 4.

IOW think we need to make portOpts.Trunk a *bool instead of a bool.

Reply:
We are not inheriting instance level setting. As discussed with @dulek above, each port makes its own decision, regardless of what the instance says.
for (1) and (4), the check was on if a port is specified via `Portopts`, if not it is set to the instance level https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1673261708882c975a746bd0c2bdd5b62bca673a/pkg/cloud/services/compute/instance.go#L178

Is it so that there are other ways of specifying multiple ports ?

Reply:
However, I think we previously *did* inherit the instance setting. Let's take a specific example:

```
      trunk: true
      ports:
      - description: "primary 1"
        foo: bar
      - description: "primary 2"
        foo: baz
```

Without this change, both of these ports will be trunked. This change will cause both of the above ports to now be trunked because `trunk` is not specified, and being a `bool` it will default to `false`. IOW this change causes a change in previously defined behaviour. I think that would be a regression.

I think the fix is simple: if we make `trunk` on the port a `*bool` instead of a `bool` we can define a third behaviour for the undefined case which is compatible with the previous behaviour.

Reply:
Thanks. Now, I understood the consequence of not making it a pointer. 
Making changes to address it. Will also update the PR description accordingly.

Reply:
Done.

Reply:
Thanks @mdbooth, I missed this detail.